script/compiled.go

```diff
@@ -40,14 +56,42 @@ func (c *Compiled) RunContext(ctx context.Context) (err error) {
 	return
 }
 
+// Clone creates a new copy of Compiled.
+// Cloned copies are safe for concurrent use by multiple goroutines.
+func (c *Compiled) Clone() *Compiled {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	clone := &Compiled{
+		globalIndexes:    c.globalIndexes,
+		bytecode:         c.bytecode,
+		globals:          make([]*objects.Object, len(c.globals)),
+		builtinFunctions: c.builtinFunctions,
```

Comment:
Is there a need to also copy/clone the `builtinFunctions` and `builtinModules`, or are they "thread-safe"?

Reply:
All states are read-only by VM, except for globals.

Reply:
What if they are being mutated by the calling host?

I'm thinking about a usecase where one makes a `compiled.Clone()` and then modifies the functions (or something to that extent).

Reply:
Ah, looks like one can't mutate them from outside `Compiled`, only when compiling the `Script`, so once you get a `Compiled` instance, it's all final at that point... ;)

Reply:
Yeah, I know that's possible technically. But, in a way, all Compiled instances are considered to be outcome of the same Script, which had defined all globals, builtins, and, etc. So, conceptually, those states are defined/managed by Script, not by Compiled. And, if one wants to change those states, it's encouraged to create a new Script instance.

Reply:
Makes sense... as long as it's ok to mutate the global variables, the functions, etc... should stay static.  The best way, to be honest, would be to enable support for defining functions in the script with arguments, and be able to invoke those functions from the host by passing the function arguments externally... 😉

Reply:
Can you elaborate?

In most of my use cases, I actually force the user script to `export` a function and load it as a module from a wrapper code. Look at this example: https://pbr.cloud/simulator _(remember this is one of work-in-progress projects)_.

Reply:
My example would be something like this. 
Imagine you have a user/tengo script loaded from disk that has a function defined, something simple like:
```
Handle := func(req, resp) {
    return {req: req, resp: resp}
}
```
In the host, we load the script and then try to invoke the `Handle(Request, Response)` function, but by passing the `Request` and `Response` objects from the host:
```
func main() {
    req := "Imagine an HTTP Request struct"
    resp := "Imagine an HTTP Response struct"
    s := script.New([]byte(code))
    c, _ := s.Compile()
    if _, err := c.RunWithArgs("Handle", req, resp); err != nil {
        panic(err)
    }
}
```

Reply:
Yeah, that's a frequent usage pattern. Calling a function (defined in global scope) directly from the host can be useful _(for another PR)_. But, regardlessly, the arguments need to be simple values (that can be automatically converted to existing primitive types), or, it needs a wrapper that implements `objects.Object` interface.

Here's an example HTTP request handling context object that I use in another project:

```golang
package server

import (
	"bytes"
	"io/ioutil"
	"net/http"

	"github.com/d5/tengo/compiler/token"
	"github.com/d5/tengo/objects"
)

type ExecContext struct {
	reqHeader   *objects.ImmutableMap
	queryParams *objects.ImmutableMap
	reqBody     *objects.Bytes
	status      int64
	resHeader   map[string]string
	resBody     bytes.Buffer
}

func NewExecContext(req *http.Request) (*ExecContext, error) {
	// request header
	reqHeader := make(map[string]objects.Object)
	for key, values := range req.Header {
		if len(values) > 0 {
			reqHeader[key] = &objects.String{Value: values[0]}
		}
	}

	// request query params
	queryParams := make(map[string]objects.Object)
	for name, values := range req.URL.Query() {
		if name != "s" && len(values) > 0 {
			queryParams[name] = &objects.String{Value: values[0]}
		}
	}

	// request body
	reqBody, err := ioutil.ReadAll(req.Body)
	if err != nil {
		return nil, err
	}

	return &ExecContext{
		reqHeader:   &objects.ImmutableMap{Value: reqHeader},
		queryParams: &objects.ImmutableMap{Value: queryParams},
		reqBody:     &objects.Bytes{Value: reqBody},
		status:      http.StatusOK,
		resHeader:   make(map[string]string),
	}, nil
}

func (c *ExecContext) TypeName() string {
	return "context"
}

func (c *ExecContext) String() string {
	return "<context>"
}

func (c *ExecContext) BinaryOp(op token.Token, rhs objects.Object) (objects.Object, error) {
	return nil, objects.ErrInvalidOperator
}

func (c *ExecContext) IsFalsy() bool {
	return false
}

func (c *ExecContext) Equals(another objects.Object) bool {
	return false
}

func (c *ExecContext) Copy() objects.Object {
	return c // always shared
}

func (c *ExecContext) IndexGet(index objects.Object) (objects.Object, error) {
	indexStr, ok := objects.ToString(index)
	if !ok {
		return nil, objects.ErrInvalidIndexType
	}

	switch indexStr {
	case "req_header":
		return c.reqHeader, nil
	case "req_body":
		return c.reqBody, nil
	case "query":
		return c.queryParams, nil
	case "set_status":
		return &objects.UserFunction{Value: c.setStatus}, nil
	case "set_header":
		return &objects.UserFunction{Value: c.setHeader}, nil
	case "delete_header":
		return &objects.UserFunction{Value: c.deleteHeader}, nil
	case "write":
		return &objects.UserFunction{Value: c.write}, nil
	default:
		return objects.UndefinedValue, nil
	}
}

func (c *ExecContext) Status() int {
	return int(c.status)
}

func (c *ExecContext) ResHeader() map[string]string {
	return c.resHeader
}

func (c *ExecContext) ResBody() []byte {
	return c.resBody.Bytes()
}

func (c *ExecContext) setStatus(args ...objects.Object) (ret objects.Object, err error) {
	if len(args) != 1 {
		return nil, objects.ErrWrongNumArguments
	}

	status, ok := objects.ToInt64(args[0])
	if !ok {
		return nil, objects.ErrInvalidArgumentType{
			Name:     "status",
			Expected: "int(compatible)",
			Found:    args[0].TypeName(),
		}
	}

	c.status = status

	return objects.UndefinedValue, nil
}

func (c *ExecContext) setHeader(args ...objects.Object) (ret objects.Object, err error) {
	if len(args) != 2 {
		return nil, objects.ErrWrongNumArguments
	}

	key, ok := objects.ToString(args[0])
	if !ok {
		return nil, objects.ErrInvalidArgumentType{
			Name:     "key",
			Expected: "string(compatible)",
			Found:    args[0].TypeName(),
		}
	}

	value, ok := objects.ToString(args[1])
	if !ok {
		return nil, objects.ErrInvalidArgumentType{
			Name:     "value",
			Expected: "string(compatible)",
			Found:    args[1].TypeName(),
		}
	}

	c.resHeader[key] = value

	return objects.UndefinedValue, nil
}

func (c *ExecContext) deleteHeader(args ...objects.Object) (ret objects.Object, err error) {
	if len(args) != 1 {
		return nil, objects.ErrWrongNumArguments
	}

	key, ok := objects.ToString(args[0])
	if !ok {
		return nil, objects.ErrInvalidArgumentType{
			Name:     "key",
			Expected: "string(compatible)",
			Found:    args[0].TypeName(),
		}
	}

	delete(c.resHeader, key)

	return objects.UndefinedValue, nil
}

func (c *ExecContext) write(args ...objects.Object) (ret objects.Object, err error) {
	if len(args) != 1 {
		return nil, objects.ErrWrongNumArguments
	}

	if v, ok := args[0].(*objects.Bytes); ok {
		_, err := c.resBody.Write(v.Value)
		if err != nil {
			return &objects.Error{Value: &objects.String{Value: err.Error()}}, nil
		}
	} else {
		v, ok := objects.ToString(args[0])
		if !ok {
			return nil, objects.ErrInvalidArgumentType{
				Name:     "value",
				Expected: "bytes/string(compatible)",
				Found:    args[0].TypeName(),
			}
		}
		_, err := c.resBody.WriteString(v)
		if err != nil {
			return &objects.Error{Value: &objects.String{Value: err.Error()}}, nil
		}
	}

	return objects.UndefinedValue, nil
}
```

Then, I create an instance and pass it to the script like this:

```golang
ctx := createContext(req, res) // *ExecContext which implements objects.Object
compiled.Set("ctx", ctx)
compiled.Run()

for key, value := range ctx.ResHeader() {
	res.Header().Set(key, value)
}
res.WriteHeader(ctx.Status())
_, _ = res.Write(ctx.ResBody())
```

Also the user code is loaded as a module by this code wrapper:

```golang
s := script.New([]byte(`import("script")(ctx)`)
```

And the actual user code should look like

```golang
rand := import("rand")

export func(ctx) {
    ctx.set_status(200)
    
    ctx.set_header("Content-Type", "text/plain")
    
    ctx.write(string(rand.intn(100)))
}
```

Reply:
Makes sense... I guess it's just not as "elegant"...  Also forces you to do `compiled.Clone()` for concurrent executions, as the variable `ctx` gets bound to global scope, instead of the method stack, so makes things a bit complicated. 😉